form-flow-valtimo/src/main/kotlin/com/ritense/valtimo/formflow/FormFlowProcessLinkTaskProvider.kt

```diff
@@ -16,25 +16,107 @@
 
 package com.ritense.valtimo.formflow
 
+import com.ritense.document.exception.DocumentNotFoundException
+import com.ritense.document.service.DocumentService
+import com.ritense.formflow.domain.definition.FormFlowDefinition
+import com.ritense.formflow.domain.definition.FormFlowDefinitionId
+import com.ritense.formflow.domain.instance.FormFlowInstance
 import com.ritense.formflow.service.FormFlowService
+import com.ritense.formlink.domain.FormAssociation
 import com.ritense.formlink.domain.FormLink
 import com.ritense.formlink.domain.ProcessLinkTaskProvider
 import com.ritense.formlink.domain.TaskOpenResult
 import com.ritense.formlink.domain.impl.formassociation.formlink.BpmnElementFormFlowIdLink
+import com.ritense.formlink.service.FormAssociationService
+import org.apache.logging.log4j.util.Strings
+import org.camunda.bpm.engine.RepositoryService
+import org.camunda.bpm.engine.RuntimeService
 import org.camunda.bpm.engine.task.Task
 
 class FormFlowProcessLinkTaskProvider(
-    val formFlowService: FormFlowService
+    private val formFlowService: FormFlowService,
+    private val formAssociationService: FormAssociationService,
+    private val documentService: DocumentService,
+    private val repositoryService: RepositoryService,
+    private val runtimeService: RuntimeService,
 ): ProcessLinkTaskProvider<FormFlowTaskOpenResultProperties> {
-    private val FORM_FLOW_TASK_TYPE_KEY = "form-flow"
 
     override fun supports(formLink: FormLink?): Boolean {
         return formLink is BpmnElementFormFlowIdLink
     }
 
     override fun getTaskResult(task: Task, formLink: FormLink): TaskOpenResult<FormFlowTaskOpenResultProperties> {
         val instances = formFlowService.findInstances(mapOf("taskInstanceId" to task.id))
-        assert(instances.size == 1)
-        return TaskOpenResult(FORM_FLOW_TASK_TYPE_KEY, FormFlowTaskOpenResultProperties(instances[0].id.id))
+        val instance = when (instances.size) {
+            0 -> createFormFlowInstance(task)
+            1 -> instances[0]
+            else -> throw IllegalStateException("Multiple form flow instances linked to task: ${task.id}")
+        }
+        return TaskOpenResult(FORM_FLOW_TASK_TYPE_KEY, FormFlowTaskOpenResultProperties(instance.id.id))
     }
-}
\ No newline at end of file
+
+    private fun createFormFlowInstance(task: Task): FormFlowInstance {
+        val additionalProperties: MutableMap<String, Any> = getAdditionalProperties(task)
+        val formFlowId = getFormAssociationByTask(task).formLink.formFlowId
+        val formFlowDefinition: FormFlowDefinition = getFormFlowDefinition(formFlowId)
+        return formFlowService.save(formFlowDefinition.createInstance(additionalProperties))
```

Comment:
It really bugs me that we pass a MutableMap for no good reason. Can't we change the signature of it like this?:
`    fun createInstance(additionalProperties: Map<String, Any>) : FormFlowInstance {
        return FormFlowInstance(formFlowDefinition = this,
            additionalProperties = additionalProperties.toMutableMap())
    }`

It won't be breaking, because a MutableMap is also a Map.

Reply:
changed